File: library/src/main/java/me/proxer/library/entity/info/InfoTag.java
package me.proxer.library.entity.info;

import com.squareup.moshi.Json;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.Value;
import me.proxer.library.api.NumberBasedBoolean;
import me.proxer.library.entity.ProxerDateItem;
import me.proxer.library.entity.ProxerIdItem;

import javax.annotation.Nullable;
import java.util.Date;

/**
 * Entity representing a single tag in the context of an {@link Entry}.
 *
 * @author Ruben Gees
 */
@Value
@EqualsAndHashCode(onParam = @__({@Nullable}))
public class InfoTag implements ProxerIdItem, ProxerDateItem {

    /**
     * Returns the id of the associated entry.
     */
    @Getter(onMethod = @__({@Override}))
    @Json(name = "tid")
    private String id;

    /**
     * Returns the id of the associated tag entry.
     */
    @Json(name = "id")
    private String entryTagId;

    /**
     * Returns the time.
     */
    @Getter(onMethod = @__({@Override}))
    @Json(name = "timestamp")
    private Date date;

    /**
     * Returns true, if this tag has been rated.
     */
    @NumberBasedBoolean
    @Json(name = "rate_flag")
    private boolean rated;

    /**
     * Returns true, if this tag is a spoiler.
     */
    @NumberBasedBoolean
    @Json(name = "spoiler_flag")
    private boolean spoiler;

    /**
     * Returns the name.
     */
    @Json(name = "tag")
    private String name;

    /**
     * Returns the description.
     */
    @Json(name = "description")
    private String description;
}

File: library/src/main/kotlin/me/proxer/library/entity/info/InfoTag.kt
package me.proxer.library.entity.info

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass
import me.proxer.library.api.NumberBasedBoolean
import me.proxer.library.entity.ProxerDateItem
import me.proxer.library.entity.ProxerIdItem
import java.util.Date

/**
 * Entity representing a single tag in the context of an [Entry].
 *
 * @property entryTagId The id of the associated tag entry.
 * @property isRated If this tag has been rated.
 * @property isSpoiler If this tag is a spoiler.
 * @property name The name.
 * @property description The description.
 *
 * @author Ruben Gees
 */
@JsonClass(generateAdapter = true)
data class InfoTag(
    @Json(name = "id") override val id: String,
    @Json(name = "tid") val entryTagId: String,
    @Json(name = "timestamp") override val date: Date,
    @field:NumberBasedBoolean @Json(name = "rate_flag") val isRated: Boolean,
    @field:NumberBasedBoolean @Json(name = "spoiler_flag") val isSpoiler: Boolean,
    @Json(name = "tag") val name: String,
    @Json(name = "description") val description: String
) : ProxerIdItem, ProxerDateItem
